### PR TITLE
[Nonlinear] add support for atan2

### LINKED ISSUE #1986: binary atan function is not supported natively
Based on discussion [here](https://discourse.julialang.org/t/how-to-save-nonlinear-jump-model-to-file/86322/18).

There is no binary version of ```atan``` function in nonlinear subsystem of MOI.

If user uses binary ```atan``` in JuMP it registers a user defined function which is not supported with ```write_to_file``` and also makes unary version unavailable.

Binary version of ```atan``` commonly [known](https://en.wikipedia.org/wiki/Atan2) as ```atan2``` is widely used in many nonlinear problems. It is supported in almost all language's math libraries and most nonlinear optimization tools.

The best solution would be to support both unary and binary versions of ```atan``` simultaneously within MOI.

Workaround would be to add separate ```atan2``` function. The drawback of such approach would be incompatibility with ```julia``` expressions witch use binary version of ```atan```. For example when extracting expressions from functions with ```Symbolics.jl```.

If we have to stick with only one version of ```atan``` perhaps the binary should be used instead of unary. As ```atan(x) = atan2(x, 1)``` so unary is easily implemented with binary. The only issue then is some extra computation ```atan2``` has over regular ```atan```.

The same issue could be inspected for some other multimethod functions in julia.

### PR DESCRIPTION
**Pull request by odow**
Closes #1986

I think this is another win for the work that has gone into nonlinear. It was pretty trivial to add support for a new multivariate function.

cc @a6a3uh

### CODE REVIEW

**Comment by odow** [comment]
Oops. The Julia function is  `atan(y, x)`. I have things backward.

**Comment by odow** [comment]
```Julia
julia> using JuMP, Ipopt

julia> model = Model(Ipopt.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Ipopt

julia> @variable(model, x >= 0.1)
x

julia> @variable(model, y >= 0)
y

julia> @objective(model, Max, y)
y

julia> @constraint(model, x^2 + y^2 <= 1)
x² + y² ≤ 1.0

julia> @NLconstraint(model, atan(y, x) <= π / 4)
(atan(y, x) - 3.141592653589793 / 4.0) - 0.0 ≤ 0

julia> optimize!(model)
This is Ipopt version 3.14.4, running with linear solver MUMPS 5.4.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        4
Number of nonzeros in Lagrangian Hessian.............:        5

Total number of variables............................:        2
                     variables with only lower bounds:        2
                variables with lower and upper bounds:        0
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        2
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        2

iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
   0  9.9999900e-03 0.00e+00 9.16e-01  -1.0 0.00e+00    -  0.00e+00 0.00e+00   0
   1  2.5702742e-02 0.00e+00 2.90e-01  -1.7 1.23e-01    -  7.56e-01 1.00e+00f  1
   2  2.1732187e-01 0.00e+00 2.57e-01  -1.7 3.43e+00    -  2.24e-02 1.70e-01f  1
   3  5.4698892e-01 3.00e-01 1.74e-01  -1.7 1.18e+00    -  1.00e+00 6.85e-01h  1
   4  7.2996883e-01 5.90e-02 3.07e-02  -1.7 2.75e-01    -  1.00e+00 1.00e+00h  1
   5  7.0251711e-01 0.00e+00 1.74e-03  -2.5 4.19e-02    -  1.00e+00 1.00e+00h  1
   6  7.0677354e-01 0.00e+00 1.95e-05  -3.8 7.73e-03    -  1.00e+00 1.00e+00h  1
   7  7.0710304e-01 0.00e+00 6.47e-08  -5.7 5.07e-04    -  1.00e+00 1.00e+00h  1
   8  7.0710679e-01 0.00e+00 9.35e-12  -8.6 5.43e-06    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 8

                                   (scaled)                 (unscaled)
Objective...............:  -7.0710678677185734e-01    7.0710678677185734e-01
Dual infeasibility......:   9.3510606726917504e-12    9.3510606726917504e-12
Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
Variable bound violation:   0.0000000000000000e+00    0.0000000000000000e+00
Complementarity.........:   2.5165591405122160e-09    2.5165591405122160e-09
Overall NLP error.......:   2.5165591405122160e-09    2.5165591405122160e-09


Number of objective function evaluations             = 9
Number of objective gradient evaluations             = 9
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 9
Number of equality constraint Jacobian evaluations   = 0
Number of inequality constraint Jacobian evaluations = 9
Number of Lagrangian Hessian evaluations             = 8
Total seconds in IPOPT                               = 0.005

EXIT: Optimal Solution Found.

julia> value(x), value(y)
(0.7071067776491126, 0.7071067867718573)
```

**Comment by odow** [comment]
@a6a3uh try this branch and see if it works with some of your models. `] add MathOptInterface#od/atan2`

**Comment by odow** [comment]
Confirmed this worked with AmplNLWriter also.

File: src/FileFormats/NL/NLExpr.jl [hunk 1]
```diff
@@ -62,7 +62,9 @@ const _JULIA_TO_AMPL = Dict{Symbol,Int}(
     :cosh => OP_cosh,
     :cos => OP_cos,
     :atanh => OP_atanh,
-    # OP_atan2 = 48,
+    # Because :atan is also a univariate operator, the is special cased in
```
Comment (odow) [inline src/FileFormats/NL/NLExpr.jl:65]:
```suggestion
    # Because :atan is also a univariate operator, this is special cased in
```

**Comment by a6a3uh** [comment]
> @a6a3uh try this branch and see if it works with some of your models. `] add MathOptInterface#od/atan2`

Thank you @odow for quick fix! I've tested the branch and it seems working. At least it produces no errors and able to calculate some trajectories for me.

**Pull request merged**